src/main/java/usdot/fhwa/stol/c2c/c2c_mvt/decoders/JsonDecoder.java

```diff
@@ -0,0 +1,182 @@
+	 * @throws C2CMVTException
+	 */
+	@Override
+	public ArrayList<JsonC2CMessage> decode(byte[] yBytes)
+		throws C2CMVTException
+	{
+		if (!checkSecurity(yBytes))
+			throw new C2CMVTException(new Exception("Found possible security threat. Did not attempt validation."), null);
+		
+		ArrayList<byte[]> yMessages = separateMessages(yBytes);
+		
+		int nMsgCount = 0;
+		try
+		{
+			ArrayList<JsonC2CMessage> oMessages = new ArrayList(yMessages.size());
+			for (byte[] yMsg : yMessages)
+			{
+				oMessages.add(checkSyntax(yMsg));
+				++nMsgCount;
+			}
+			
+			// for testing purposing
+			{
+				for (JsonC2CMessage oMsg : oMessages)
+					StandardValidationController.addLogRecord(oMsg.getMessage());
+			} // remove this block before release
+			
+			return oMessages;
+		}
+		catch (JsonParseException oEx)
+		{
+			throw new C2CMVTException(oEx, String.format("Error occured while checking syntax of message %d of %d", nMsgCount + 1, yMessages.size()));
+		}
+	}
+
+	/**
+	 * Checks for multiple valid messages defined by the bytes. Messages can be separated
+	 * by an empty string, a single comma, or whitespace characters. Tests for
+	 * and ignores Byte order marking and ensures that the bytes could define
+	 * a JSON Object or Array by testing for matching {} or [] respectively
+	 * 
+	 * 
+	 * @param yBytes the message(s) in bytes
+	 * @return list of separate messages as byte[]
+	 * @throws C2CMVTException
+	 */
+	@Override
+	protected ArrayList<byte[]> separateMessages(byte[] yBytes)
+		throws C2CMVTException
+	{
+		ArrayList<byte[]> yMessages = new ArrayList();
+		try (ByteArrayInputStream oBaos = new ByteArrayInputStream(yBytes);
+			BufferedReader oReader = new BufferedReader(new InputStreamReader(oBaos, Charset.forName(m_sEncoding))))
+		{
+			ByteArrayOutputStream oOut = new ByteArrayOutputStream();
+			BufferedWriter oWriter = new BufferedWriter(new OutputStreamWriter(oOut, Charset.forName(m_sEncoding)));
+			if (yBytes.length > 2 && yBytes[0] == (byte)0xEF && yBytes[1] == (byte)0xBB && yBytes[2] == (byte)0xBF) // skip BOM, only allowing UTF-8 right now
+			{
+				for (int n = 0; n < 3; n++)
+					oBaos.read();
+			}
+			int nChar;
+			int nBraceCount = 1;
+			char cStartingBrace = (char)oReader.read();
+			if (cStartingBrace != '[' && cStartingBrace != '{')
+				throw new Exception("JSON data must start with [ or {");
+			oWriter.write(cStartingBrace);
+			char cClosingBrace = cStartingBrace == '[' ? ']' : '}';
+			while ((nChar = oReader.read()) >= 0)
+			{
+				char cCur = (char)nChar;
+				oWriter.write(nChar);
+				if (cCur == cStartingBrace)
+					++nBraceCount;
+				else if (cCur == cClosingBrace)
+					--nBraceCount;
+				if (nBraceCount == 0)
+				{
+					oWriter.close();
+					yMessages.add(oOut.toByteArray());
+					oOut.reset();
+					oWriter = new BufferedWriter(new OutputStreamWriter(oOut, Charset.forName(m_sEncoding)));
+					int nCommaCount = 0;
+					while ((nChar = oReader.read()) >= 0)
```

Comment:
Do we need the second while loop? If the sole difference between the first and second while loop is skipping commas and whitespace,  we can use an indicator `isFirstJSONObject`to determine whether skip whilespace and commas.

Reply:
I am rewriting this code to handle some edge cases and will try to avoid a second loop

Reply:
Updated code. No longer have a second loop